crates/bevy_ecs/src/archetype.rs

```diff
@@ -15,21 +16,36 @@ use std::{
 };
 
 #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
-pub struct ArchetypeId(usize);
+pub struct ArchetypeId(NonMaxUsize);
 
 impl ArchetypeId {
-    pub const EMPTY: ArchetypeId = ArchetypeId(0);
-    pub const RESOURCE: ArchetypeId = ArchetypeId(1);
-    pub const INVALID: ArchetypeId = ArchetypeId(usize::MAX);
+    pub const EMPTY: ArchetypeId = unsafe { ArchetypeId::new_unchecked(0) };
```

Comment:
I like this better than just plain niching. The dedicated resource archetype in particular is nice.